docs(angular): update mfe guide (#9500)
nrwl/nx: docs/shared/guides/setup-mfe-with-angular.md

[Webpack 5](https://webpack.js.org/blog/2020-10-10-webpack-5-release/) introduced a [Module Federation Plugin](https://webpack.js.org/concepts/module-federation/#modulefederationplugin-high-level) enabling multiple, independently built and deployed bundles of code to form a single application. This is the foundation of Micro Frontend Architecture and the Module Federation Plugin makes implementing such an architecture much simpler.  
With Angular 12 adding support for Webpack 5 it increases the viability of scaffolding a Micro Frontend architecture with Angular.

> We have added some generators to aid in the scaffolding of Module Federation configuration required for setting up a Micro Frontend Architecture.

Reviewer: ```suggestion
We have added some generators to aid in the scaffolding of the Module Federation configuration required for setting up a Micro Frontend Architecture.
```

Therefore, using Nx, it can be fairly straightforward to scaffold and build a Micro Frontend Architecture from a monorepo with all the additional benefits of Nx.

> In this guide, we'll show you to how setup a Micro Frontend Architecture with Nx and Angular.

Reviewer: ```suggestion
In this guide, we'll show you to how to set up a Micro Frontend Architecture with Nx and Angular.
```

### Creating our apps

We need to generate two applications that support Module Federation.

We'll start with the Admin Dashboard application which will act as a host application for the MFE:

```bash
# Npm
npx nx g @nrwl/angular:host dashboard
```

```bash
# Yarn
yarn nx g @nrwl/angular:host dashboard
```

> The application generator will create and modify the files needed to setup the Angular application.

Reviewer: ```suggestion
The application generator will create and modify the files needed to set up the Angular application.
```

Now, let's generate the Login application as a remote application.

```bash
# Npm
npx nx g @nrwl/angular:remote login --host=dashboard
```

```bash
# Yarn
yarn nx g @nrwl/angular:remote login --host=dashboard
```

> _**Note:** We provided `--host=dashboard` as an option. This tells the generator that this remote app will be consumed by the Dashboard application. The generator will automatically link these two apps together in the `mfe.config.js` that gets used in the `webpack.config.js`_

Reviewer: ```suggestion
_**Note:** We provided `--host=dashboard` as an option. This tells the generator that this remote application will be consumed by the Dashboard application. The generator will automatically link these two applications together in the `mfe.config.js` that gets used in the `webpack.config.js`._
```

> _**Note**: The `RemoteEntryModule` generated will be imported in `app.module.ts` file, however, it is not used in the `AppModule` itself. This it to allow TS to find the Module during compilation, allowing it to be included in the built bundle. This is required for the Module Federation Plugin to expose the Module correctly. You can choose to import the `RemoteEntryModule` in the `AppModule` if you wish, however, it is not necessary._

Reviewer: ```suggestion
_**Note**: The `RemoteEntryModule` generated will be imported in `app.module.ts` file, however, it is not used in the `AppModule` itself. This is to allow TS to find the Module during compilation, allowing it to be included in the built bundle. This is required for the Module Federation Plugin to expose the Module correctly. You can choose to import the `RemoteEntryModule` in the `AppModule` if you wish, however, it is not necessary._
```

For both apps, the generator did the following:

- Created the standard Angular application files
- Added a `mfe.config.js` file
- Added a `webpack.config.js` and `webpack.prod.config.js`
- Added a `bootstrap.ts` file
- Moved the code that is normally in `main.ts` to `bootstrap.ts`
- Changed `main.ts` to dynamically import `bootstrap.ts` _(this is required for the Module Federation to correct load versions of shared libraries)_

> - Updated the `build` target in the `project.json` to use the `@nrwl/angular:webpack-browser` executor _(this is required as it supports passing a custom webpack configuration to the Angular compiler)_

Reviewer: ```suggestion
- Updated the `build` target in the `project.json` to use the `@nrwl/angular:webpack-browser` executor _(this is required as it supports passing a custom Webpack configuration to the Angular compiler)_
```

> - Updated the `serve` target to use `@nrwl/angular:webpack-server`. _(this is required as we first need webpack to build the app with our custom webpack config)_

Reviewer: ```suggestion
- Updated the `serve` target to use `@nrwl/angular:webpack-server` _(this is required as we first need Webpack to build the application with our custom Webpack configuration)_
```

> The key differences reside within the configuration of the Module Federation Plugin within each app's `mfe.config.js`.

Reviewer: ```suggestion
The key differences reside within the configuration of the Module Federation Plugin within each application's `mfe.config.js`.
```

We see the following within Login's micro frontend configuration:

```js
module.exports = {
  name: 'login',
  exposes: {
    './Module': 'apps/login/src/app/remote-entry/entry.module.ts',
  },
};
```

Taking a look at each property of the configuration in turn:

> - `name` is the name that Webpack assigns to the remote appliction. It **must** match the name of the application.

Reviewer: ```suggestion
- `name` is the name that Webpack assigns to the remote application. It **must** match the name of the application.
```

Now let's create the Dashboard application where we'll hide some content if the user is not authenticated. If the user is not authenticated, we will present them with the Login application where they can log in.

> For this to work, the state within `UserService` must be shared across both applications. Usually with Module Federation in webpack, you have to specifiy the packages to share between all the apps in your Micro Frontend solution.  

Reviewer: ```suggestion
For this to work, the state within `UserService` must be shared across both applications. Usually, with Module Federation in Webpack, you have to specify the packages to share between all the applications in your Micro Frontend solution.  
```

However, by taking advantage of Nx's project graph, Nx will automatically find and share the dependencies of your applications.

_**Note:** This helps to enforce a single version policy and reduces the risk of [Micro Frontend Anarchy](https://www.thoughtworks.com/radar/techniques/micro-frontend-anarchy)_

> Now, let's delete the `app.component.html` and `app.component.css` files in the Dashboard app. They will not be needed for this tutorial.

Reviewer: ```suggestion
Now, let's delete the `app.component.html` and `app.component.css` files in the Dashboard application. They will not be needed for this tutorial.
```
